core/middlewares/loader.js

```diff
@@ -0,0 +1,162 @@
+                    output[middlewareName] = {
+                        enabled: true,
+                        order: 0,
+                        group: 'default',
+                        indexPath: pluginIndexPath
+                    };
+
+                    // Load middleware options
+                    var configPath = path.join(dest, dir, 'options.js');
+                    if (fs.existsSync(configPath)) {
+                        var middlewareConfig = require(configPath);
+
+                        _.forOwn(output[middlewareName], function (value, key) {
+                            var overVal = middlewareConfig[key];
+
+                            if (overVal) {
+                                if (key === 'order') {
+                                    if (overVal >= 0) {
+                                        output[middlewareName][key] = overVal;
+                                    } else {
+                                        log.warn('Middlewares are restricted to define order with value lower than 0 (zero). Please modify '+ middlewareName+ ' middleware options.');
+                                    }
+                                } else {
+                                    output[middlewareName][key] = overVal;
+                                }
+                            }
+                        });
+                    }
+                }
+            }
+        });
+    }
+
+    return output;
+};
+
+var sortMiddlewares = function(groupsOrder, list){
+    var output = [];
+
+    if (!_.isArray(groupsOrder) && !list) return output;
+
+    var groupedList = {};
+
+    // Sort by groups
+    _.forOwn(list, function (value, key) {
+        var group = value.group || 'default';
+        var middleware = value;
+
+        if (!middleware.enabled) return;
+
+        middleware.name = middleware.name || key;
+
+        groupedList[group] = groupedList[group] || [];
+        groupedList[group].push(value);
+    });
+
+    // Sort each group
+    _.forOwn(groupedList, function (value, key) {
+        groupedList[key] = _.sortByOrder(value, ['order'], ['asc']);
+    });
+
+    // Concat groups by order
+    groupsOrder.forEach(function(item){
+        if (groupedList[item]) output = output.concat(groupedList[item]);
+    });
+
+    return output;
+};
+
+var loadMiddlewares = function(listArr, app){
+    if (!_.isArray(listArr) && !app) return;
+
+    log.debug('loading', listArr);
+
+    listArr.forEach(function(item){
+        if (item && fs.existsSync(item.indexPath)) {
+            app.use(require(item.indexPath).process);
+        }
+    });
+};
+
+module.exports.process = function(app, globalOptions){
+    var config = {
+        loadGroupsOrder: [
+            'request',
+            'pre-html',
+            'default',
+            'html',
+            'response'
+        ],
+        list: {
+            md: {
+                enabled: true,
+                order: -1,
```

Comment:
it would be nice to use readable constants instead of magic numbers (`-1, 0, -2` in order)